### PR TITLE
SCUMM HE: Online/Network Multiplayer Support

### PR DESCRIPTION
**Pull request by LittleToonCat**
List of things left to do before this could be merged (more could be added due to feedback):
- [ ] Cleanup `net_main.cpp`, using `JSONObjects` instead of constructing messages via string formatting, like how `net_lobby.cpp` does.
- [ ] Document code.
- [ ] Put debug messages in a separate debug channel.
- [ ] Move `sessionselector.*` to SCUMM engine and bump theme version.
- [ ] Session and player name setup dialog for Backyard Football 2002.
- [ ] ENet needs a test compile on various different platforms and disabled on configure as needed.
-----
This has been a pet project I've been working on for multiple months, and I believe it's finally time to gather some code feedback. :)

This adds network multiplayer play supported by the compatible Humongous Entertainment games: Backyard Football, Backyard Baseball 2001, Backyard Football 2002, and Moonbase Commander.

It includes a new networking backend based of [ENet](http://enet.bespin.org).  It is primarily used for hosting and joining peer-to-peer game sessions and transferring in-game data.  For Internet-wide game sessions, it communicates with a separate "session server" to advertise and join such sessions, with basic hole-punching to bypass NAT, if that does not work, the server will relay data for the user.  Hosting from and joining to other devices in your local network is also possible.  It also contains lobby client code for Football 1999 and Baseball 2001. 

The ENet backend is located in `backends/networking/enet`, while the source code for its latest stable version (1.3.17) is located in the `source` directory.  While the whole netcode itself is located in `engines/scumm/he/net`

A couple of years back, I've gotten online play for Backyard Football (the 1999 version) and Backyard Baseball 2001 working on an [old and soon to be obsolete fork](https://github.com/Backyard-Sports-Online/scummvm), and support for these games is provided by [Backyard Sports Online](https://backyardsports.online), a service and community I maintain.  Clicking the register button on these games in this branch has a brief mention of the service, this is because ScummVM will be generously providing support and hosting of the project itself. (Thanks sev and rootfather!)

At Backyard Sports Online, we have our competitive online leagues, and have implemented game modifications that suit our needs, notably custom teams, which includes kids not normally playable online (such as Season Play NPCs and Mr. Clanky).  We've included such mods here, but under a toggle option in the game settings (disabled by default), since such mods are not needed for casual players.  sev has said this is completely fine.

The old lobby code communication was previously done via `SDL2_net`, but for this one, I've decided to use libcurl, simply because it can communicate with TLS/SSL servers and does the handshaking, making the connection secure.  The new backend code can be found in `backends/networking/curl/socket.cpp`.

Server code and more information for both the session and lobby servers will be located in the `scummvm-sites` repository.  The PR for the server could be found here: scummvm/scummvm-sites/pull/20

For those interested in testing: You should be able to checkout the branch from the fork and start hosting games as the default servers are already up and running as of writing, feedback is greatly appreciated.

Special Thanks:
- Vissery for testing the sports games with me.
- Autumm, iDUMB, and "Fluff?" from the Moonbase Commander Discord for Moonbase testing and support.
- PatchMixolydic, QuestionMonkey, Sarah (onfy) for miscellaneous testing.
- The Backyard Sports Online community and admins for their ongoing love and support since the very start. :heart:

### CODE REVIEW

**Comment by BLooperZ** [comment]
Amazing work, thank you!

File: engines/scumm/he/script_v72he.cpp [hunk 1]
```diff
@@ -1190,6 +1190,79 @@ void ScummEngine_v72he::o72_arrayOps() {
...
+							res = val2 & val1;
+							break;
+						case 4: // Binary OR
+							res = val2 | val1;
+							break;
+						case 5: // Binary XOR
+							res = val2 ^ val1;
+							break;
+						default:
+							error("o72_arrayOps: case 132 unknown type %d)", type);
```
[the remark this answers is not in the document]
Comment (LittleToonCat) [reply engines/scumm/he/script_v72he.cpp]: Good eye, thank you, fixed in 8294a18ee33ef8084c6037fb3aaade4763eb5dbf.

File: engines/scumm/he/net/net_lobby.h [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+	void gameStarted(int hoster, int player, int playerNameArray);
+	void gameFinished();
+
+};
+
+} // End of namespace Scumm
+
+
+#endif
```
Comment (sev-) [inline engines/scumm/he/net/net_lobby.h:190]: No end line

File: engines/scumm/he/net/net_main.cpp [hunk 3]
```diff
@@ -0,0 +1,1447 @@
...
+			startQuerySessions(false);
+			if (!_broadcastSocket) {
+				return 0;
+			}
+
+			_sessions.clear();
+			_broadcastSocket->send(address.host.c_str(), 9130, "{\"cmd\": \"get_session\"}");
+
+			uint tickCount = 0;
+			while(!_sessions.size()) {
```
Comment (sev-) [inline engines/scumm/he/net/net_main.cpp:156]: little formatting violation

File: engines/scumm/he/net/net_main.cpp [hunk 4]
```diff
@@ -0,0 +1,1447 @@
...
+				_sessionServerAddress.port = 9120;
+		}
+		if (_sessionHost->connectPeer(_sessionServerAddress.host, _sessionServerAddress.port)) {
+			// FIXME: Get the IP address of the session server when a domain address is used.
+
+			// _sessionServerPeer = _sessionHost->getPeerIndexFromHost(_sessionServerAddress.host, _sessionServerAddress.port);
+			_sessionServerPeer = 0;
+			// Create session to the session server.
+			Common::String req = Common::String::format(
+				"{\"cmd\":\"host_session\",\"game\":\"%s\",\"version\":\"%s\",\"name\":\"%s\",\"maxplayers\":%d}",
```
Comment (sev-) [inline engines/scumm/he/net/net_main.cpp:299]: I wonder, is it absolutely necessary to use a JSON object here as you mention in your TODO? I think, the code will go like 5x with JSON.

File: engines/scumm/he/net/net_main.cpp [hunk 5]
```diff
@@ -0,0 +1,1447 @@
...
+	}
+	return address;
+}
+
+Common::String Net::getStringFromAddress(Address address) {
+	return Common::String::format("%s:%d", address.host.c_str(), address.port);
+}
+
+void Net::setSessionServer(Common::String sessionServer) {
+	debug(1, "Net::setSessionServer(\"%s\")", sessionServer.c_str());
```
Comment (sev-) [inline engines/scumm/he/net/net_main.cpp:100]: I would consider putting all of these into a separate debug channel

File: engines/scumm/he/net/net_main.cpp [hunk 6]
```diff
@@ -0,0 +1,1447 @@
...
+	if (_isHost && _hostDataQueue.size()) {
+		_isShuttingDown = true;
+		// Send out any remaining data from the queue before shutting down.
+		while (_hostDataQueue.size()) {
+			if (_hostDataQueue.size() != _hostDataQueue.size())
+				warning("NETWORK: Sizes of data and peer index queues does not match!  Expect some wonky stuff");
+			Common::JSONValue *json = _hostDataQueue.pop();
+			int peerIndex = _peerIndexQueue.pop();
+			handleGameDataHost(json, peerIndex);
+		_isShuttingDown = false;
```
Comment (sev-) [inline engines/scumm/he/net/net_main.cpp:435]: Indentation is off

File: engines/scumm/he/net/net_main.cpp [hunk 7]
```diff
@@ -0,0 +1,1447 @@
...
+	// char name[MAX_PROVIDER_NAME];
+	// char ipaddress[MAX_IP_SIZE];
+
+	// ipaddress[0] = '\0';
+
+	// _vm->getStringFromArray(parameter1, name, sizeof(name));
+	// if (parameter2)
+	// 	_vm->getStringFromArray(parameter2, ipaddress, sizeof(ipaddress));
+
+	// debug(1, "Net::setProviderByName(\"%s\", \"%s\")", name, ipaddress); // PN_SetProviderByName
```
Comment (sev-) [inline engines/scumm/he/net/net_main.cpp:514]: Perhaps, you still want to uncomment it for completeness' sake.

File: engines/scumm/he/net/net_main.cpp [hunk 8]
```diff
@@ -0,0 +1,1447 @@
...
+void Net::getProviderName(int providerIndex, char *buffer, int length) {
+	warning("STUB: Net::getProviderName(%d, \"%s\", %d)", providerIndex, buffer, length); // PN_GetProviderName
+}
+
+void Net::serviceSessionServer() {
+	if (!_sessionServerHost)
+		return;
+
+	uint8 type = _sessionServerHost->service();
+	switch(type) {
```
Comment (sev-) [inline engines/scumm/he/net/net_main.cpp:883]: formatting

File: engines/scumm/he/net/net_main.cpp [hunk 9]
```diff
@@ -0,0 +1,1447 @@
...
+			if (root.contains("address") && root.contains("sessions")) {
+				_hostPort = getAddressFromString(root["address"]->asString()).port;
+				Common::JSONArray sessions = root["sessions"]->asArray();
+				for (uint i = 0; i != sessions.size(); i++) {
+					Common::JSONObject sessionData = sessions[i]->asObject();
+					Address sessionAddress = getAddressFromString(sessionData["address"]->asString());
+
+					// Check if we already know about this session:
+					bool makeNewSession = true;
+					for (Common::Array<Session>::iterator j = _sessions.begin(); j != _sessions.end(); j++) {
```
Comment (sev-) [inline engines/scumm/he/net/net_main.cpp:928]: `for (auto &j : _sessions)`

File: engines/scumm/he/net/net_main.cpp [hunk 10]
```diff
@@ -0,0 +1,1447 @@
...
+					// Game/Version mismatch.
+					return;
+
+				if (players < 1 || players > _maxPlayers)
+					// This session is either full or not finished initalizing (adding the host player itself)
+					return;
+
+				// Check if the session of the game ID (from the internet session server) exists.
+				// if so, update it as a local session and swap the internet-based address to local.
+				for (Common::Array<Session>::iterator i = _sessions.begin(); i != _sessions.end(); i++) {
```
Comment (sev-) [inline engines/scumm/he/net/net_main.cpp:1079]: You could use iterative for here as well

File: engines/scumm/he/net/net_main.cpp [hunk 11]
```diff
@@ -0,0 +1,1447 @@
...
+	uint8 messageType = _sessionHost->service();
+	switch (messageType) {
+	case ENET_EVENT_TYPE_NONE:
+		return true;
+	case ENET_EVENT_TYPE_CONNECT:
+		{
+			debug(1, "NETWORK: New connection from %s:%d", _sessionHost->getHost().c_str(), _sessionHost->getPort());
+			return true;
+		}
+		return true;
```
Comment (sev-) [inline engines/scumm/he/net/net_main.cpp:1122]: This code is unreachable

File: engines/scumm/he/net/net_main.cpp [hunk 12]
```diff
@@ -0,0 +1,1447 @@
...
+				// TODO: Host migration
+				if (!_isHost && _vm->_currentRoom == 2) {
+					_vm->displayMessage(0, "You have been disconnected from the host.\nNormally, host migration would take place, but ScummVM doesn't do that yet, so this game session will now end." );
+					_vm->VAR(253) = 26; // gGameMode = GAME-OVER
+					_vm->runScript(2104, 1, 0, 0); // leave-game
+				}
+			}
+			return true;
+		}
+		return true;
```
Comment (sev-) [inline engines/scumm/he/net/net_main.cpp:1145]: This code is unreachable

File: engines/scumm/he/net/net_main.cpp [hunk 13]
```diff
@@ -0,0 +1,1447 @@
...
+						handleGameData(json, peerIndex);
+				}
+			}
+			if (_sessionHost)
+				_sessionHost->destroyPacket();
+		}
+		return true;
+		break;
+	}
+	return true;
```
Comment (sev-) [inline engines/scumm/he/net/net_main.cpp:1236]: This code is unreachable

File: engines/scumm/he/net/net_main.cpp [hunk 14]
```diff
@@ -0,0 +1,1447 @@
...
+						handleGameDataHost(json, peerIndex);
+					else
+						handleGameData(json, peerIndex);
+				}
+			}
+			if (_sessionHost)
+				_sessionHost->destroyPacket();
+		}
+		return true;
+		break;
```
Comment (sev-) [inline engines/scumm/he/net/net_main.cpp:1234]: This code is unreachable

File: engines/scumm/he/net/net_main.cpp [hunk 15]
```diff
@@ -0,0 +1,1447 @@
...
+				}
+			}
+		}
+		break;
+	default:
+		warning("NETWORK: Unknown data type: %d", to);
+	}
+}
+
+} // End of namespace Scumm
```
Comment (sev-) [inline engines/scumm/he/net/net_main.cpp:1447]: Missing end of line

File: engines/scumm/he/net/net_main.h [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+	Address _sessionServerAddress;
+	bool _forcedAddress;
+	bool _gotSessions;
+	int _sessionServerPeer;
+	bool _isRelayingGame; // If we're relaying in-game data over the session server or not.
+};
+
+} // End of namespace Scumm
+
+#endif
```
Comment (sev-) [inline engines/scumm/he/net/net_main.h:174]: Missing end of the line. I think there must be some setting in your editor to keep these.

File: engines/scumm/script.cpp [hunk 17]
```diff
@@ -558,6 +558,17 @@ int ScummEngine::readVar(uint var) {
 			return !ConfMan.getBool("subtitles");
 		}
 
+#if defined(USE_ENET) && defined(USE_LIBCURL)
+		if (ConfMan.getBool("enable_competitive_mods")) {
+			// HACK: If we're reading var586, competitive mods enabled, playing online,
+			// successfully fetched custom teams, and we're not in one of the three scripts
+			// that cause bugs if 263 is returned here, return 263.
+			if ( _game.id == GID_BASEBALL2001 && var == 586 && readVar(399) == 1 && readVar(747) == 1 &&
```
Comment (sev-) [inline engines/scumm/script.cpp:566]: formatting `if (_game.id...`

File: engines/scumm/script.cpp [hunk 18]
```diff
@@ -558,6 +558,17 @@ int ScummEngine::readVar(uint var) {
...
+				!(_currentRoom == 4 && ( vm.slot[_currentScript].number == 2150 || vm.slot[_currentScript].number == 2208 || vm.slot[_currentScript].number == 2210))) {
```
Comment (sev-) [inline engines/scumm/script.cpp:567]: Formatting `(vm.slot...`

File: gui/sessionselector.cpp [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+ *
+ */
+
+#include "engines/metaengine.h"
+#include "common/algorithm.h"
+#include "common/config-manager.h"
+#include "common/debug.h"
+#include "common/system.h"
+#include "common/taskbar.h"
+#include "common/translation.h"
```
Comment (sev-) [inline gui/sessionselector.cpp:28]: You also have to add this file to `po/POTFILES`, so it goes to our Weblate for translation.

File: gui/sessionselector.cpp [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+SessionSelectorDialog::SessionSelectorDialog(Scumm::ScummEngine_v90he *vm)
+	: Dialog("SessionSelector"),
+	_vm(vm),
+	_joinButton(nullptr),
+	_queryProgressText(nullptr) {
+
+	_timestamp = 0;
+
+	_queryProgressText = new StaticTextWidget(this, "SessionSelector.QueryProgressText",
+						_("Querying games..."));
```
Comment (sev-) [inline gui/sessionselector.cpp:50]: Please add a commend for translators here with `I18N: ` prefix. Something like
```C++
// I18N: Retrieving list of online games
```

File: gui/sessionselector.cpp [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+
+		for (int i = 0; i < numSessions; i++) {
+			char name[MAX_SESSION_NAME];
+			_vm->_net->getSessionName(i, name, MAX_SESSION_NAME);
+			_list->append(name);
+		}
+
+		_joinButton->setEnabled(false);
+		// Update the dialog
+		_queryProgressText->setLabel(Common::U32String::format(_("Found %d available games."), l.size()));
```
Comment (sev-) [inline gui/sessionselector.cpp:107]: Here is also, please add a comment to the translators that the games here are online and active ones, not like game files.

File: gui/sessionselector.h [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#ifndef SESSION_SELECTOR_DIALOG_H
```
Comment (sev-) [inline gui/sessionselector.h:22]: Please rename it to match the current file name

File: gui/sessionselector.h [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+#define SESSION_SELECTOR_DIALOG_H
+
+#include "gui/dialog.h"
+#include "gui/widgets/list.h"
+#include "common/fs.h"
+#include "common/hashmap.h"
+#include "common/stack.h"
+#include "common/str.h"
+
+#include "scumm/he/intern_he.h"
```
Comment (sev-) [inline gui/sessionselector.h:32]: Oops. You cannot do it from the GUI. This means that you have to move these two files, `sessionselector.*` back to `engines/scumm`. Perhaps, call it there `dialog_sessionselector.h`

**Review by sev- — CHANGES_REQUESTED** [review]
Did my first round of reviews. `sessionselector.*` has to be moved back to the SCUMM engine, since gui/ code cannot depend on some engine code.

Also, please bump themes version (in THEMERC in every theme dir as well as in `gui/ThemeEngine.h` and put themes regeneration, e.g. .zip files into a separate commit.

File: base/commandLine.cpp [hunk 24]
```diff
@@ -213,6 +213,14 @@ static const char HELP_STRING4[] =
 #ifdef ENABLE_SCUMM
 	"  --tempo=NUM              Set music tempo (in percent, 50-200) for SCUMM games\n"
 	"                           (default: 100)\n"
+#endif
+#if defined(ENABLE_HE) && defined(USE_ENET)
+	"  --host-game              Host an online game for Moonbase Commander.\n"
```
[the remark this answers is not in the document]
Comment (LittleToonCat) [reply base/commandLine.cpp:218]: Yes, this overrides the "hostip" config in `o80_readConfigFile` which gets read by the game at startup.  This is implemented, so I could get testing done on the game faster without having to go through the main menu.  Same for the `--join-game=IP` argument right below it, which does the same but for "joinip".

File: engines/scumm/he/net/net_main.h (hunk 16 above)
Comment (LittleToonCat) [reply engines/scumm/he/net/net_main.h]: Just enabled the "Insert Final Newline" setting in Visual Studio Code, which I didn't know existed until just now.  This shouldn't be a problem anymore, thank you.

**Comment by sev-** [comment]
Thank you for addressing all my notes.

Once it is rebased, so there are no conflicts, I'll merge and the work could be continued in-tree

**Comment by sev-** [comment]
Thank you!

**Pull request merged**